File: A2/python-service/src/db/connection.py
import asyncio
import asyncpg
from config.config import Config
import os

_pool = None


async def get_pool():
    """Get or create database connection pool (Bulkhead pattern)"""
    global _pool
    if _pool is None:
        config = Config()
        db_config = config.get_db_config()
        
        # Bulkhead: Create isolated connection pool for database
        _pool = await asyncpg.create_pool(
            host=db_config['host'],
            port=db_config['port'],
            database=db_config['database'],
            user=db_config['user'],
            password=db_config['password'],
            min_size=db_config['min_size'],
            max_size=db_config['max_size'],
        )
        print('Database connection pool established')
    return _pool


async def initialize_database():
    """Initialize database with migrations"""
    pool = await get_pool()
    
    try:
        # Read and execute migration files
        migrations_dir = os.path.join(os.path.dirname(__file__), '../migrations')
        
        with open(os.path.join(migrations_dir, '001_create_sensors.sql'), 'r') as f:
            migration1 = f.read()
            async with pool.acquire() as conn:
                await conn.execute(migration1)
        
        with open(os.path.join(migrations_dir, '002_seed_data.sql'), 'r') as f:
            migration2 = f.read()
            async with pool.acquire() as conn:
                await conn.execute(migration2)
        
        print('Database initialized successfully')
    except Exception as error:
        print(f'Database initialization failed: {error}')
        raise


async def close_pool():
    """Close database connection pool"""
    global _pool
    if _pool:
        await _pool.close()
        _pool = None

File: A4/services/node-sensor-service/src/db/connection.js
const { Pool } = require('pg');
const Config = require('../config/config');

let pool = null;

function getPool() {
  if (!pool) {
    const config = new Config();
    const dbConfig = config.getDbConfig();
    
    // Bulkhead: Create isolated connection pool for database
    pool = new Pool(dbConfig);
    
    // Handle pool errors gracefully
    pool.on('error', (err) => {
      console.error('Unexpected error on idle client', err);
    });
    
    // Test connection on startup
    pool.query('SELECT 1')
      .then(() => {
        console.log('Database connection pool established');
      })
      .catch((err) => {
        console.error('Failed to establish database connection pool', err);
      });
  }
  
  return pool;
}

async function initializeDatabase() {
  const pool = getPool();
  const fs = require('fs').promises;
  const path = require('path');
  
  try {
    // Read and execute migration files
    const migration1 = await fs.readFile(
      path.join(__dirname, '../migrations/001_create_sensors.sql'),
      'utf8'
    );
    await pool.query(migration1);
    
    const migration2 = await fs.readFile(
      path.join(__dirname, '../migrations/002_seed_data.sql'),
      'utf8'
    );
    await pool.query(migration2);
    
    console.log('Database initialized successfully');
  } catch (error) {
    console.error('Database initialization failed:', error);
    throw error;
  }
}

async function closePool() {
  if (pool) {
    await pool.end();
    pool = null;
  }
}

module.exports = {
  getPool,
  initializeDatabase,
  closePool
};
